File: conformance/tests/p384-primitives/p384-primitives.runar.ts
import { SmartContract, assert, p384Add, p384Mul, p384MulGen, p384OnCurve } from 'runar-lang';
import type { P384Point } from 'runar-lang';

class P384Primitives extends SmartContract {
  readonly expectedPoint: P384Point;

  constructor(expectedPoint: P384Point) {
    super(expectedPoint);
    this.expectedPoint = expectedPoint;
  }

  public verify(k: bigint, basePoint: P384Point) {
    const result: P384Point = p384Mul(basePoint, k);
    assert(p384OnCurve(result));
    assert(result == this.expectedPoint);
  }

  public verifyAdd(a: P384Point, b: P384Point) {
    const result: P384Point = p384Add(a, b);
    assert(p384OnCurve(result));
    assert(result == this.expectedPoint);
  }

  public verifyMulGen(k: bigint) {
    const result: P384Point = p384MulGen(k);
    assert(p384OnCurve(result));
    assert(result == this.expectedPoint);
  }
}

File: examples/java/src/main/java/runar/examples/p384-primitives/P384Primitives.runar.java
package runar.examples.p384primitives;

import java.math.BigInteger;

import runar.lang.SmartContract;
import runar.lang.annotations.Public;
import runar.lang.annotations.Readonly;
import runar.lang.types.P384Point;

import static runar.lang.Builtins.assertThat;
import static runar.lang.Builtins.p384Add;
import static runar.lang.Builtins.p384Mul;
import static runar.lang.Builtins.p384MulGen;
import static runar.lang.Builtins.p384OnCurve;

/**
 * P384Primitives -- conformance fixture for the NIST P-384 (secp384r1)
 * EC primitives. Mirrors {@link runar.examples.p256primitives.P256Primitives}
 * but at the larger 96-byte point size.
 */
class P384Primitives extends SmartContract {

    @Readonly P384Point expectedPoint;

    P384Primitives(P384Point expectedPoint) {
        super(expectedPoint);
        this.expectedPoint = expectedPoint;
    }

    @Public
    void verify(BigInteger k, P384Point basePoint) {
        P384Point result = p384Mul(basePoint, k);
        assertThat(p384OnCurve(result));
        assertThat(result.equals(this.expectedPoint));
    }

    @Public
    void verifyAdd(P384Point a, P384Point b) {
        P384Point result = p384Add(a, b);
        assertThat(p384OnCurve(result));
        assertThat(result.equals(this.expectedPoint));
    }

    @Public
    void verifyMulGen(BigInteger k) {
        P384Point result = p384MulGen(k);
        assertThat(p384OnCurve(result));
        assertThat(result.equals(this.expectedPoint));
    }
}
